src/components/Views/CameraView/CameraView.js

```diff
@@ -101,87 +129,239 @@ class CameraView extends React.Component<
         const data = await this.camera.takePictureAsync(options);
         this.setState({ loading: false });
 
-        if (selectedObservation) {
-          updateObservation({
-            ...selectedObservation,
-            media: selectedObservation.media.concat([
-              {
-                type: 'Photo',
-                source: data.uri
-              }
-            ])
+        if (fromEditor) {
+          if (selectedObservation) {
+            updateObservation({
+              ...selectedObservation,
+              media: selectedObservation.media.concat([
+                {
+                  type: 'Photo',
+                  source: data.uri
+                }
+              ])
+            });
+            goToObservationEditor();
+          }
+        } else {
+          goToCategories();
+          const initialObservation = applyObservationDefaults({
+            id: size(observations) + 1,
+            createdFrom: 'camera'
           });
-          goToObservationEditor();
+          createObservation(initialObservation);
+
+          navigator.geolocation.getCurrentPosition(
+            position => {
+              const { latitude, longitude } = position.coords;
+
+              updateObservation({
+                ...(selectedObservation || initialObservation),
+                lat: Math.round(latitude * 1000) / 1000,
+                lon: Math.round(longitude * 1000) / 1000,
+                media: initialObservation.media.concat([
+                  {
+                    type: 'Photo',
+                    source: data.uri
+                  }
+                ])
+              });
+            },
+            error => console.warn(error)
+          );
         }
       } catch (error) {
         console.warn(error);
       }
     }
   };
 
+  closeRightDrawer = () => {
+    this.rightDrawer.close();
+  };
+
+  openRightDrawer = () => {
+    this.rightDrawer.open();
+  };
+
+  handleRightDrawerRef = (ref: Drawer) => {
+    this.rightDrawer = ref;
+  };
+
+  goToMapView = () => {
+    this.props.navigation.navigate({ routeName: 'MapView' });
```

Comment:
let's do this through a dispatchProp. I'd prefer if this goes through our actions process so it's synchronous.

Reply:
👍